src/mpz.rs

```diff
@@ -158,31 +165,22 @@ impl Mpz {
     //        machinery for a custom type.
     pub fn to_str_radix(&self, base: u8) -> String {
         unsafe {
+            assert!(base >= 2 && base <= 36, "invalid base");
             // Extra two bytes are for possible minus sign and null terminator
-            let len = __gmpz_sizeinbase(&self.mpz, base as c_int) as usize + 2;
+            let len = {
+                let len = __gmpz_sizeinbase(&self.mpz, base as c_int) as usize;
+                assert!(usize::MAX - len >= 2, "capacity overflow");
+                len + 2
+            };
 
             // Allocate and write into a raw *c_char of the correct length
             let mut vector: Vec<u8> = Vec::with_capacity(len);
-            vector.set_len(len);
-
             __gmpz_get_str(vector.as_mut_ptr() as *mut _, base as c_int, &self.mpz);
-
-            let mut first_nul = None;
-            let mut index : usize = 0;
-            for elem in &vector {
-                if *elem == 0 {
-                    first_nul = Some(index);
-                    break;
-                }
-                index += 1;
-            }
-            let first_nul = first_nul.unwrap_or(len);
-
-            vector.truncate(first_nul);
-            match String::from_utf8(vector) {
-                Ok(s)  => s,
-                Err(_) => panic!("GMP returned invalid UTF-8!")
-            }
+            let string_len = strnlen(vector.as_ptr(), len);
```

Comment:
Excuse my ignorance; is the memory pointed to by Rust vectors guaranteed to be null terminated? I did a quick search but couldn't find docs on this, thanks in advance for the lesson :)

Reply:
It is not, but my understanding is that `__gmpz_get_str` NUL-terminates the buffer it writes to.  I could be wrong, though; I wrote this code around two years ago.